monero-woocommerce-gateway.php

```diff
@@ -0,0 +1,249 @@
+<?php
+/*
+Plugin Name: Monero Woocommerce Gateway
+Plugin URI: https://github.com/monero-integrations/monerowp
+Description: Extends WooCommerce by adding a Monero Gateway
+Version: 3.0.0
+Tested up to: 4.9.8
+Author: mosu-forge, SerHack
+Author URI: https://monerointegrations.com/
+*/
+// This code isn't for Dark Net Markets, please report them to Authority!
+
+defined( 'ABSPATH' ) || exit;
+
+// Constants, you can edit these if you fork this repo
+define('MONERO_GATEWAY_MAINNET_EXPLORER_URL', 'https://xmrchain.net/');
+define('MONERO_GATEWAY_TESTNET_EXPLORER_URL', 'https://testnet.xmrchain.com/');
+define('MONERO_GATEWAY_ADDRESS_PREFIX', 0x12);
+define('MONERO_GATEWAY_ADDRESS_PREFIX_INTEGRATED', 0x13);
+define('MONERO_GATEWAY_ATOMIC_UNITS', 12);
+define('MONERO_GATEWAY_ATOMIC_UNIT_THRESHOLD', 10); // Amount under in atomic units payment is valid
+define('MONERO_GATEWAY_DIFFICULTY_TARGET', 120);
+
+// Do not edit these constants
+define('MONERO_GATEWAY_PLUGIN_DIR', plugin_dir_path(__FILE__));
+define('MONERO_GATEWAY_PLUGIN_URL', plugin_dir_url(__FILE__));
+define('MONERO_GATEWAY_ATOMIC_UNITS_POW', pow(10, MONERO_GATEWAY_ATOMIC_UNITS));
+define('MONERO_GATEWAY_ATOMIC_UNITS_SPRINTF', '%.'.MONERO_GATEWAY_ATOMIC_UNITS.'f');
+
+// Include our Gateway Class and register Payment Gateway with WooCommerce
+add_action('plugins_loaded', 'monero_init', 1);
+function monero_init() {
+
+    // If the class doesn't exist (== WooCommerce isn't installed), return NULL
+    if (!class_exists('WC_Payment_Gateway')) return;
+
+    // If we made it this far, then include our Gateway Class
+    require_once('include/class-monero-gateway.php');
+
+    // Create a new instance of the gateway so we have static variables set up
+    new Monero_Gateway($add_action=false);
+
+    // Include our Admin interface class
+    require_once('include/admin/class-monero-admin-interface.php');
+
+    add_filter('woocommerce_payment_gateways', 'monero_gateway');
+    function monero_gateway($methods) {
+        $methods[] = 'Monero_Gateway';
+        return $methods;
+    }
+
+    add_filter('plugin_action_links_' . plugin_basename(__FILE__), 'monero_payment');
+    function monero_payment($links) {
+        $plugin_links = array(
+            '<a href="'.admin_url('admin.php?page=monero_gateway_settings').'">'.__('Settings', 'monero_gateway').'</a>'
+        );
+        return array_merge($plugin_links, $links);
+    }
+
+    add_filter('cron_schedules', 'monero_cron_add_one_minute');
+    function monero_cron_add_one_minute($schedules) {
+        $schedules['one_minute'] = array(
+            'interval' => 60,
+            'display' => __('Once every minute', 'monero_gateway')
+        );
+        return $schedules;
+    }
+
+    add_action('wp', 'monero_activate_cron');
+    function monero_activate_cron() {
+        if(!wp_next_scheduled('monero_update_event')) {
+            wp_schedule_event(time(), 'one_minute', 'monero_update_event');
```

Comment:
Waste of resources.

Reply:
The problem with the way the current plugin works is that it only does payment checking logic when the order confirmation page is refreshed. So if a user leaves the order confirmation page, then their order will never confirm. Cron is necessary for allowing users to leave the page, and still have their order confirmed, especially if a merchant sets confirms to 10+. If you feel once a minute is too often, that can be changed.

Reply:
Imagine having running 200 cronjobs at the same time. I still disagree for running cronjobs when users leaves the order confirmation page.

Reply:
To be clear, the plugin only runs one cronjob for all payments. It does two things:  
* Fetches the price so live exchange rates can be displayed on the site, or used for displaying prices in XMR on the frontend  
* Checks for pending orders, and if there are any, checks for payments

Reply:
I know this PR is closed, but for the time that you re-look at this code I wanted to add:

Right now the current plugin actually runs the same payment checking code that this cron job does every 17 seconds, so this code should actually be more efficient.

Reply:
This, or something like this, is critical to include indeed. It is unacceptable that the page cannot be closed or the order will not fulfill.